File: CopyListWithRandomPointer.py
# Definition for singly-linked list with a random pointer.
class RandomListNode:
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        if not head:
            return None
        p = head
        # Copy every node
        while p:
            tmpNode = RandomListNode(p.label)
            tmpNode.next = p.next
            p.next = tmpNode
            p = p.next.next

        # Copy random info
        p = head
        while p:
            if p.random:
                p.next.random = p.random.next
            p = p.next.next

        # Detach the linkedlist into two
        copyhead = head.next
        pold, pnew = head, copyhead
        while pnew.next:
            pold.next = pnew.next
            pold = pold.next
            pnew.next = pold.next
            pnew = pnew.next
        
        pold.next = None
        pnew.next = None
        
        return copyhead
        

File: CPP/Second/copy_list_with_random_pointer.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        if (!head)
            return nullptr;
        RandomListNode *old = head, *nw;
        // copy linkedlist's each node
        while(old) {
            nw = new RandomListNode(old->label);
            nw->next = old->next;
            old->next = nw;
            old = nw->next;
        }
        // copy random pointers
        old = head;
        while(old) {
            nw = old->next;
            // some node's random is none
            if (old->random)
                nw->random = old->random->next;
            old = nw->next;
        }
        // detach two lists
        RandomListNode *new_head = head->next;
        old = head;
        while (old) {
            nw = old->next;
            old->next = nw->next;
            // in case of the last one
            if (nw->next)
                nw->next = nw->next->next;
            old = old->next;
        }
        return new_head;
    }
};


int main(int argc, char *argv[])
{

    return 0;
}
